### ISSUE TITLE
Nested searches

### ISSUE THREAD

**Issue by sqs** [sourcegraph/sourcegraph#1005]
**NOTE:** See https://github.com/sourcegraph/sourcegraph/issues/1005#issuecomment-473011076 for the latest update on this issue.

Also known as subquery search. This is about improving the power of search in
Sourcegraph via adding sub-queries and logical operators (AND, OR, NOT). These
sort of queries are not uncommon via the shell. Usually done by mixing shell
pipelines, `find`, `grep -L` and `grep -l`. Examples of things you can achieve:

* Make it possible to find all repositories which do not contain a Dockerfile.
* Make it possible to find results which match X, but only if the repository contains the word Y.
* Make it possible to find results which match X, but only if the file contains the word Y.

Here are some example queries in our proposed syntax to achieve the above:

* `type:repo f:Dockerfile`
* `type:repo -f:Dockerfile`
* `(type:repo langserver) cxp`
* `(type:file Copyright) Redistribution`
* `(type:repo (file:^license or file:readme) BSD) (type:repo (type:file -f:Dockerfile)) test`

Alternate syntax proposed in https://github.com/sourcegraph/sourcegraph/issues/4774

### Roadmap

- [x] support type:repo and type:file in zoekt - [done](https://github.com/sourcegraph/zoekt/commit/f3a07d825c6a9231ec4fafbe8b4d0564abe4323e)
- [x] search providers - [done](https://github.com/sourcegraph/sourcegraph/commit/dfddd7278103a109d91add9008897d8216064d0b#diff-fd05f777fb54d29e95adbbc78cf7aaa6)
- [x] nested queries in searcher - [done](https://github.com/sourcegraph/sourcegraph/commit/8065e2c2bd34866f8561e467f7aac27d6a59e9c4#diff-3874e864697e5acfad0a70edabe58a50)
- [ ] split search into pkgs for code coverage stats #3566 
- [ ] add integration test harness of search in a single process #3475 
- [ ] add integration tests for all customer requested nested search functionality #3531
- [ ] fix any existing breakage found by the new integration tests #3532
- [ ] performance testing and tuning of searcher
   - [ ] add endpoint showing top recent queries #3256
   - [ ] add tool or endpoint to generate a report comparing performance of search with and without nesting enabled for non-indexed search #2879
   - [ ] make non-indexed search perform as well with nesting enabled as without #3497
- [ ] use the nested query parser for all queries, translating them to old-style query structures behind the scenes #3495
- [ ] use the data obtained from #3256 to ensure most common queries give the same results as before
- [ ] feature parity for repo queries (forked, archive, repogroups)
- [ ] backwards compatibility with diff search
- [ ] update UI to use new syntax (eg how to specify refs )
- [ ] update sample (saved) queries 
- [ ] update documentation

Closes:

- #110 

Customers:

- https://app.hubspot.com/contacts/2762526/company/419771425 
- https://app.hubspot.com/contacts/2762526/company/732945850 
- https://app.hubspot.com/contacts/2762526/company/464956351  
- https://app.hubspot.com/contacts/2762526/company/407948923

**Comment by keegancsmith** [comment]
The backwards compatibility has been the big slowdown, especially with respect to the way resolvers interact in the graphql layer. I'm busy hacking in an approach which merges the two approaches (old and new), just so we can ship this for text search and then incrementally support other backends.

**Comment by keegancsmith** [comment]
I've updated the issue description with the latest status (see top note) and plan (see added bullet points). There are lots of bullet points to go, but they are relatively minor compared to the work done so far.

**Comment by nicksnyder** [comment]
This is still important, but not within our strict definition of what is required for 3.0, so I move this to 3.1 milestone.

**Comment by hho** [comment]
> Please comment on this issue any bugs you may find.

When adding `count:1000` to the end of a sub-query (as happens when you press the "show more" button at the bottom of the results list), no results are shown at all.

**Comment by ijt** [comment]
I'm grabbing this issue based on my understanding that I'm owning this now.

**Comment by slimsag** [comment]
@ijt @nicksnyder @sqs I think this issue and https://github.com/sourcegraph/sourcegraph/issues/35  will be important for us long term. No timeline in mind, but probably important for us.

If we can get them placed into a milestone other than `Backlog` so we have a general idea of when we think they may be done (e.g. 6mo from now or something), that would be cool. Then if we need them sooner we can dedicate more resources to them.

**Comment by nicksnyder** [comment]
@ijt and @ijsnow are going to start working on a design document for this from a clean slate. We won't put this into a milestone until we are reasonable confident that we know when it will ship, but the timeline we have communicated to @ijsnow and @ijt is 3-6 month timeframe. We are prioritizing quality and testing over shipping this sooner.

**Comment by slimsag** [comment]
Great, that's perfect!

**Comment by slimsag** [comment]
@ijt @ijsnow @nicksnyder can someone update this issue to reflect our current thinking? Is this still accurate?

> the timeline we have communicated to @ijsnow and @ijt is 3-6 month timeframe. We are prioritizing quality and testing over shipping this sooner.

**Comment by nicksnyder** [comment]
It is up to @ijt and @ijsnow to spec out what work needs to be done prior to actually starting to work on nested search (please do this). To a first approximation this work potentially includes (but is not limited to):
- Customer issues with our existing search
    - https://github.com/sourcegraph/sourcegraph/issues/3294
    - https://github.com/sourcegraph/sourcegraph/issues/3057
    - https://github.com/sourcegraph/sourcegraph/issues/2970
- Adding appropriate metrics so we can track baseline search performance and then compare it to nested search perf so we know that enabling nested search won't cause per regressions.
    - https://github.com/sourcegraph/sourcegraph/issues/3256
    - https://github.com/sourcegraph/sourcegraph/issues/2879
- Prospective improvements to our architecture that might necessary to have a good user experience with search performance
    - https://github.com/sourcegraph/sourcegraph/issues/3261

**Comment by ijt** [comment]
@nicksnyder based on yesterday's conversation it looks like we are going to hold off on #3261 for now and focus on getting nested querying to work first.

**Comment by nicksnyder** [comment]
This feature is backlogged until we solve the problem of making Sourcegraph search easier for novice users, which is what we are prioritizing thinking about now.

When we revisit nested searches, we will favor solutions that are novice friendly.

**Comment by dobrou** [comment]
Sad to hear this feature is postponed :(

Maybe following proposal could be considered novice friendly and could be taken from backlog sooner than later :)

What about extending idea of new operator 'repohasfile:' and introduce new operator 'filehastext:'

**A new search filter _filehastext:"regexp"_ would allow users to filter results to just files containing a matching regexp.**
Including it's negative variant  _-filehastext:"regexp"_ to exclude such files.
For example:
_"class Helper" filehastext:"namespace org,my.company"_ 
would show search results for Helper classes but only in in files that mention "namespace org,my.company".

**Comment by olafurpg** [comment]
Can this issue be closed? Sourcegraph now has search expressions (`AND`, `OR`) and filters (`repo:contains.file(foo)`). Docs https://docs.google.com/document/d/1_m63fsBMAtqaq3GA_aMzKUPxD3yxTy8d12lJE6qN6PU/edit#

cc/ @rvantonder

**Comment by lguychard** [comment]
Yes, this can be closed.
